File: leetcode/dota2_senate/dota2_senate.cpp
class Solution {
public:
    string predictPartyVictory(string senate) {
        queue<int> radiants;
        queue<int> dires;
        int n=senate.size();
        for(int i=0;i<senate.size();i++){
            char senator= senate[i];
            if(senator=='R'){
                radiants.push(i);
            }else{
                dires.push(i);
            }
        }

        while(!radiants.empty()&&!dires.empty()){
            int radiant_idx = radiants.front(); radiants.pop();
            int dire_idx = dires.front(); dires.pop();

            if(radiant_idx>dire_idx){
                dires.push(dire_idx+n);
            }else{
                radiants.push(radiant_idx+n);
            }
        }

        return radiants.empty() ? "Dire" : "Radiant";
    }
};

File: leetcode/dota2_senate/dota2_senate.py
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        n=len(senate)
        dire_idxs = deque()
        radiant_idxs = deque()

        for idx, senator in enumerate(senate):
            if senator=="D":
                dire_idxs.append(idx)
            elif senator=="R":
                radiant_idxs.append(idx)
        
        while dire_idxs and radiant_idxs:
            dire_idx = dire_idxs.popleft()
            radiant_idx = radiant_idxs.popleft()

            if dire_idx>radiant_idx:
                radiant_idxs.append(radiant_idx+n)
            else:
                dire_idxs.append(dire_idx+n)
        
        if len(dire_idxs)==0:
            return "Radiant"
        elif len(radiant_idxs)==0:
            return "Dire"
        else:
            return "what????"
